### PR TITLE
feat: create kms key as part of cluster bootstrap

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

- Create KMS key as part of cluster boostrap
  - This is the first step to making the kms key version mandatory for the kms plugin pod.
  - This will also allow BYOK configuration with KMS with the next KMS release.
  - Deterministic behavior instead of relying on runtime for key creation.
- Fixes using `premium` sku with user-assigned managed identity

**Issue Fixed**:

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [x] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

cc @ritazh

### CODE REVIEW

**Comment by aramase** [comment]
CI failures are because of cloud-init payload limit:
```
$ az deployment group create --name kubernetes-southeastasia-63191 --resource-group kubernetes-southeastasia-63191 --template-file /__w/1/s/_output/kubernetes-southeastasia-63191/azuredeploy.json --parameters /__w/1/s/_output/kubernetes-southeastasia-63191/azuredeploy.parameters.json
..........2021/01/16 00:19:51 
Error from deployment for kubernetes-southeastasia-63191 in resource group kubernetes-southeastasia-63191:exit status 1
2021/01/16 00:19:51 Command Output: ValidationError: Deployment failed. Correlation ID: 959b9fbd-41f9-4e00-a67f-7218e9a6f9b2. {
  "error": {
    "code": "InvalidParameter",
    "message": "Custom data in OSProfile must be in Base64 encoding and with a maximum length of 87380 characters.",
    "target": "customData"
  }
}
```

File: parts/k8s/cloud-init/artifacts/cse_main.sh [hunk 1]
```diff
@@ -219,6 +219,13 @@ time_metric "EnsureContainerd" ensureContainerd
 time_metric "EnsureDHCPv6" ensureDHCPv6
 {{end}}
 
+if [[ -n ${MASTER_NODE} ]]; then
+  {{/* configure and enable kms plugin */}}
+  {{- if EnableEncryptionWithExternalKms}}
```
Comment (jackfrancis) [inline parts/k8s/cloud-init/artifacts/cse_main.sh:224]: nit: we can save a little cloud-init overhead by moving the `{{- if EnableEncryptionWithExternalKms}}` up the conditional hierarchy (basically we don't need to even bother with this particular `if [[ -n ${MASTER_NODE} ]]; then` part if we don't have a kms config)

File: pkg/engine/masterarmresources.go [hunk 2]
```diff
@@ -192,9 +193,11 @@ func createKubernetesMasterResourcesVMSS(cs *api.ContainerService) []interface{}
 	}
 
 	if isKMSEnabled {
+		// TODO (aramase) remove storage account creation as part of kms plugin v0.0.11
```
Comment (jackfrancis) [inline pkg/engine/masterarmresources.go:196]: should we do this in this PR?

Comment (aramase) [reply pkg/engine/masterarmresources.go:196]: The current release `v0.0.10` is still dependent on the storage account to exist. I'll remove it when we cut a kms release `v0.0.11` and update the version in aks-engine.

File: parts/k8s/cloud-init/artifacts/cse_main.sh (hunk 1 above)
Comment (aramase) [reply parts/k8s/cloud-init/artifacts/cse_main.sh]: Updated!

**Review by jackfrancis — DISMISSED** [review]
/lgtm

File: .pipelines/pr-e2e.yaml [hunk 3]
```diff
@@ -64,7 +64,7 @@ jobs:
     k8sRelease: '1.18'
     apimodel: 'examples/e2e-tests/kubernetes/release/default/definition.json'
     createVNET: true
-    enableKMSEncryption: true
+    enableKMSEncryption: false
```
Comment (ritazh) [inline .pipelines/pr-e2e.yaml:67]: We don't want to test this as part of PR gate anymore?

Comment (jackfrancis) [reply .pipelines/pr-e2e.yaml:67]: The new implementation brings excess cloud-init data overhead, which exceeds the allowed (64k bytes) limit when incorporated into the cluster config we're testing.

So we've added a discrete KMS-enabled cluster config to the set of jenkins-enabled cluster configs we test out of band.

File: parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+SERVICE_PRINCIPAL_CLIENT_SECRET=$(jq -r '.aadClientSecret' ${AZURE_JSON_PATH})
+TENANT_ID=$(jq -r '.tenantId' ${AZURE_JSON_PATH})
+KMS_KEYVAULT_NAME=$(jq -r '.providerVaultName' ${AZURE_JSON_PATH})
+KMS_KEY_NAME=$(jq -r '.providerKeyName' ${AZURE_JSON_PATH})
+USER_ASSIGNED_IDENTITY_ID=$(jq -r '.userAssignedIdentityID' ${AZURE_JSON_PATH})
+PROVIDER_KEY_VERSION=$(jq -r '.providerKeyVersion' ${AZURE_JSON_PATH})
+
+TOKEN_URL="https://login.microsoftonline.com/$TENANT_ID/oauth2/token"
+SCOPE="https://vault.azure.net"
+KEYVAULT_URL="https://$KMS_KEYVAULT_NAME.vault.azure.net/keys/$KMS_KEY_NAME/versions?maxresults=1&api-version=7.1"
```
Comment (ritazh) [inline parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:17]: All three urls here will be different for diff clouds. https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L89

**Comment by aramase** [comment]
Tried to get the key version by using `outputs` and `reference` in arm-template but `reference` can't be used with `variables` or `parameters`. 

Example arm snippet used:
```json
    "resources": [
        {
            "type": "Microsoft.KeyVault/vaults/keys",
            "apiVersion": "2019-09-01",
            "name": "[concat(parameters('vaultName'), '/', parameters('keyName'))]",
            "location": "southcentralus",
            "properties": {
                "kty": "[parameters('keyType')]",
                "keyOps": "[parameters('keyOps')]",
                "keySize": "[parameters('keySize')]",
                "curveName": "[parameters('curveName')]"
            }
        }
    ],
    "variables": {
        "kmsEncryptionKey": "[resourceId('Microsoft.KeyVault/vaults/keys', parameters('vaultName'), parameters('keyName'))]"
    },
    "outputs": {
        "kmsKey": {
            "type": "string",
            "value": "[reference(resourceId('Microsoft.KeyVault/vaults/keys', parameters('vaultName'), parameters('keyName'))).keyUriWithVersion]"
        }
    }
```

https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-outputs?tabs=azure-powershell
https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource?tabs=json#reference

File: parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+set +x
+set -euo pipefail
+
+AZURE_JSON_PATH="/etc/kubernetes/azure.json"
+SERVICE_PRINCIPAL_CLIENT_ID=$(jq -r '.aadClientId' ${AZURE_JSON_PATH})
+SERVICE_PRINCIPAL_CLIENT_SECRET=$(jq -r '.aadClientSecret' ${AZURE_JSON_PATH})
+TENANT_ID=$(jq -r '.tenantId' ${AZURE_JSON_PATH})
+KMS_KEYVAULT_NAME=$(jq -r '.providerVaultName' ${AZURE_JSON_PATH})
+KMS_KEY_NAME=$(jq -r '.providerKeyName' ${AZURE_JSON_PATH})
```
Comment (ritazh) [inline parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:11]: where do you set the value name and the key name in azure.json?

Comment (aramase) [reply parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:11]: These 2 values are already set when azure.json is created the first time - https://github.com/Azure/aks-engine/blob/master/parts/k8s/cloud-init/artifacts/cse_config.sh#L212-L215

File: parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh (hunk 4 above)
Comment (aramase) [reply parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh]: Updated the script to configure url based on cloud environment

**Comment by aramase** [comment]
@jackfrancis The `pr-e2e (k8s_1_19_containerd_e2e)` failure seems to be a flake. This PR is ready for final review and merge.

@ritazh Addressed comments. PTAL!

File: parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh (hunk 5 above)
Comment (ritazh) [reply parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:11]: I was referring to how the keyname is passed into the cse_config.sh. But I see you are hardcoding it in the ARM template:
`[concat(variables('clusterKeyVaultName'), '/', 'k8s')]",`

File: .pipelines/pr-e2e.yaml (hunk 3 above)
Comment (ritazh) [reply .pipelines/pr-e2e.yaml:67]: Let's document how to run the kms-enabled jenkins job.

**Review by ritazh — APPROVED** [review]

**Pull request merged**